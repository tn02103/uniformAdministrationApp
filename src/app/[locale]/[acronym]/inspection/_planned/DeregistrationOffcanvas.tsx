import { AutocompleteField } from "@/components/fields/AutocompleteField";
import { updateCadetRegistrationForInspection } from "@/dal/inspection";
import { usePlannedInspectionList } from "@/dataFetcher/inspection";
import { useScopedI18n } from "@/lib/locales/client";
import { CadetLabel } from "@/types/globalCadetTypes";
import { PlannedInspectionType } from "@/types/inspectionTypes";
import { faX } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { format } from "date-fns";
import { useMemo } from "react";
import { Button, Offcanvas, Table } from "react-bootstrap";
import { toast } from "react-toastify";

type DeregistrationOffcanvasProps = {
    inspection: PlannedInspectionType;
    cadetList: CadetLabel[];
    onClose: () => void;
}
export const DeregistrationOffcanvas = ({ inspection, cadetList, onClose }: DeregistrationOffcanvasProps) => {
    const { mutate } = usePlannedInspectionList();
    const t = useScopedI18n('inspection.planned');

    const options = useMemo(() => {
        return cadetList.filter(c =>
            !inspection.deregistrations.some(d => d.fk_cadet === c.id)
        ).map(c => ({
            value: c.id, label: `${c.firstname} ${c.lastname}`
        })).sort((a, b) => a.label.localeCompare(b.label));
    }, [cadetList, inspection.deregistrations]);

    const handleAdd = (value: string | null) => {
        if (value) {
            const cadet = cadetList.find(c => c.id === value);
            if (!cadet) return;

            updateCadetRegistrationForInspection({
                cadetId: value,
                inspectionId: inspection.id,
                deregister: true
            }).then(() => {
                mutate();
            }).catch(() => {
                toast.error(t('errors.deregistration', cadet));
            });
        }
    }

    const handleRemove = (cadet: CadetLabel) => {
        updateCadetRegistrationForInspection({
            cadetId: cadet.id,
            inspectionId: inspection.id,
            deregister: false
        }).then(() => {
            mutate();
        }).catch(() => {
            toast.error(t('errors.register', cadet));
        });
    }

    return (
        <Offcanvas
            show
            backdrop={false}
            onHide={onClose}
            placement="end"
            scroll={true}
            style={{ width: "450px" }}
            aria-labelledby="offcanvas-inspection-deregistration"
        >
            <Offcanvas.Header closeButton>
                <Offcanvas.Title id="offcanvas-inspection-deregistration">
                    {t('deregistration.header', { name: inspection.name })}
                </Offcanvas.Title>
            </Offcanvas.Header>
            <Offcanvas.Body>
                <div className="row justify-content-between mb-3">
                    <div className="col-auto">
                        <AutocompleteField
                            resetOnChange
                            label={t('deregistration.label.add')}
                            options={options}
                            onChange={handleAdd}
                        />
                    </div>
                </div>
                <Table>
                    <thead>
                        <tr>
                            <th></th>
                            <th>{t('deregistration.label.person')}</th>
                            <th>{t('deregistration.label.date')}</th>
                        </tr>
                    </thead>
                    <tbody data-testid="deregistration-table-body" role="rowgroup" aria-label="body">
                        {inspection.deregistrations.map((dereg) => (
                            <tr key={dereg.fk_cadet} className="hoverCol">
                                <td>
                                    <Button
                                        variant={"light"}
                                        className="hoverColHidden"
                                        size="sm"
                                        onClick={() => handleRemove(dereg.cadet)}
                                        aria-label={t('deregistration.label.remove')}
                                    >
                                        <FontAwesomeIcon icon={faX} className="text-danger" />
                                    </Button>
                                </td>
                                <td>{dereg.cadet.firstname} {dereg.cadet.lastname}</td>
                                <td>{format(dereg.date, 'dd.MM.yyyy HH:mm')}</td>
                            </tr>
                        ))}
                    </tbody>
                </Table>
            </Offcanvas.Body>
        </Offcanvas >
    );
}
