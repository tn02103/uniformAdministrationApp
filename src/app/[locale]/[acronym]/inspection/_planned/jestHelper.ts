import dayjs from "@/lib/dayjs";
import { PlannedInspectionType } from "@/types/inspectionTypes";


export const mockCadetList = [
    { id: "1", firstname: "Jack", lastname: "Doe" },
    { id: "2", firstname: "Jane", lastname: "Smith" },
    { id: "3", firstname: "Alice", lastname: "Johnson" },
    { id: "4", firstname: "Bob", lastname: "Brown" },
    { id: "5", firstname: "Charlie", lastname: "Davis" },
];

export const mockInspectionList: PlannedInspectionType[] = [
    {
        id: "1777e18f-f4fa-4ec4-90d8-58a9b1ce16ac",
        name: "Inspection 1",
        date: dayjs().toDate(),
        timeStart: null,
        timeEnd: null,
        deregistrations: [
            {
                fk_cadet: mockCadetList[0].id,
                cadet: mockCadetList[0],
                date: dayjs().toDate(),
                fk_inspection: "1777e18f-f4fa-4ec4-90d8-58a9b1ce16ac",
            },
            {
                fk_cadet: mockCadetList[1].id,
                cadet: mockCadetList[1],
                date: dayjs().subtract(10, "day").toDate(),
                fk_inspection: "1777e18f-f4fa-4ec4-90d8-58a9b1ce16ac",
            }
        ]
    },
    {
        id: "1777e18f-f4fa-4ec4-90d8-58a9b1ce16ad",
        name: "Inspection 2",
        date: dayjs().add(2, "day").toDate(),
        timeStart: null,
        timeEnd: null,
        deregistrations: [
            {
                fk_cadet: mockCadetList[2].id,
                cadet: mockCadetList[2],
                date: dayjs().toDate(),
                fk_inspection: "1777e18f-f4fa-4ec4-90d8-58a9b1ce16ad",
            }
        ]
    },
    {
        id: "1777e18f-f4fa-4ec4-90d8-58a9b1ce16ae",
        name: "Inspection 3",
        date: new Date("2023-10-01"),
        timeStart: new Date("2023-10-01T10:00:00"),
        timeEnd: null,
        deregistrations: []
    }
];

jest.mock("@/dal/inspection", () => ({
    createInspection: jest.fn().mockResolvedValue("createInspection"),
    deleteInspection: jest.fn().mockResolvedValue("deleteInspection"),
    updatePlannedInspection: jest.fn().mockResolvedValue("updateInspection"),
    updateCadetRegistrationForInspection: jest.fn().mockResolvedValue("updateCadetRegistrationForInspection"),

    startInspection: jest.fn().mockResolvedValue("startInspection"),
    stopInspection: jest.fn().mockResolvedValue("stopInspection"),
}));
jest.mock("@/dataFetcher/inspection", () => ({
    usePlannedInspectionList: jest.fn().mockReturnValue({
        data: mockInspectionList,
        mutate: jest.fn().mockResolvedValue(mockInspectionList),
    }),
    useInspectionState: jest.fn().mockReturnValue({
        data: { active: false },
        mutate: jest.fn().mockResolvedValue({ active: false }),
    }),
}))
