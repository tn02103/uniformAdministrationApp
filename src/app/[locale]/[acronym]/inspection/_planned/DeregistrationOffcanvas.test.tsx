
import "./jestHelper";

import { getAllByRole, getByRole, getByTestId, queryByRole, render, screen } from "@testing-library/react";
import { mockCadetList, mockInspectionList } from "./jestHelper";
import { DeregistrationOffcanvas } from "./DeregistrationOffcanvas";
import userEvent from "@testing-library/user-event";
import { format } from "date-fns";

describe('<DeregistrationOffcanvas/>', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });
    it('should render', () => {
        const mockInspection = mockInspectionList[0];
        render(
            <DeregistrationOffcanvas
                inspection={mockInspection}
                cadetList={mockCadetList}
                onClose={jest.fn()}
            />
        );

        expect(screen.getByText(/deregistration.header/i)).toBeInTheDocument();

        // validate the table
        const table = screen.getByRole("table");
        const tbody = getByRole(table, "rowgroup", { name: /body/i });
        getAllByRole(tbody, "row").forEach((row, index) => {
            const cells = getAllByRole(row, "cell");
            const deregistration = mockInspection.deregistrations[index];

            expect(row).toHaveClass(/hoverCol/);
            expect(cells).toHaveLength(3);
            expect(getByRole(cells[0], "button")).toHaveAccessibleName(/label.remove/i);
            expect(getByRole(cells[0], "button")).toHaveClass(/hoverColHidden/);
            expect(cells[1]).toHaveTextContent(deregistration.cadet.firstname);
            expect(cells[1]).toHaveTextContent(deregistration.cadet.lastname);
            expect(cells[2]).toHaveTextContent(format(deregistration.date, "dd.MM.yyyy HH:mm"))
        });
        expect(screen.getByTestId(/autocomplete-field-group/)).toBeInTheDocument();
    });

    it('should close the offcanvas with close button', async () => {
        const user = userEvent.setup();
        const onClose = jest.fn();
        render(
            <DeregistrationOffcanvas
                inspection={mockInspectionList[0]}
                cadetList={mockCadetList}
                onClose={onClose}
            />
        );

        const closeButton = screen.getByRole("button", { name: /close/i });
        expect(closeButton).toBeInTheDocument();
        await user.click(closeButton);
        expect(onClose).toHaveBeenCalled();
    });

    it('should be able to deregister a cadet', async () => {
        const { updateCadetRegistrationForInspection } = jest.requireMock("@/dal/inspection");
        const { mutate } = jest.requireMock("@/dataFetcher/inspection").usePlannedInspectionList();

        const user = userEvent.setup();
        render(
            <DeregistrationOffcanvas
                inspection={mockInspectionList[0]}
                cadetList={mockCadetList}
                onClose={jest.fn()}
            />
        );

        const deregisterFieldGroup = screen.getByTestId("autocomplete-field-group");
        const deregisterField = getByRole(deregisterFieldGroup, "textbox");
        expect(deregisterField).toBeInTheDocument();
        expect(deregisterField).toHaveValue("");
        expect(deregisterField).toHaveAttribute("aria-expanded", "false");
        await user.click(deregisterField);
        expect(deregisterField).toHaveAttribute("aria-expanded", "true");

        const optionListBox = getByRole(deregisterFieldGroup, "listbox");
        expect(optionListBox).toBeInTheDocument();
        expect(getAllByRole(optionListBox, "option")).toHaveLength(mockCadetList.length - 2);
        expect(queryByRole(optionListBox, "option", { name: RegExp(mockCadetList[0].firstname) })).toBeNull();
        expect(queryByRole(optionListBox, "option", { name: RegExp(mockCadetList[1].firstname) })).toBeNull();
        expect(getByRole(optionListBox, "option", { name: RegExp(mockCadetList[2].firstname) })).toBeInTheDocument();
        expect(getByRole(optionListBox, "option", { name: RegExp(mockCadetList[3].firstname) })).toBeInTheDocument();
        expect(getByRole(optionListBox, "option", { name: RegExp(mockCadetList[4].firstname) })).toBeInTheDocument();

        await user.type(deregisterField, "o");
        expect(deregisterField).toHaveValue("o");
        expect(getAllByRole(optionListBox, "option")).toHaveLength(2);
        expect(getByRole(optionListBox, "option", { name: RegExp(mockCadetList[2].firstname) })).toBeInTheDocument();
        expect(getByRole(optionListBox, "option", { name: RegExp(mockCadetList[3].firstname) })).toBeInTheDocument();
        expect(queryByRole(optionListBox, "option", { name: RegExp(mockCadetList[4].firstname) })).toBeNull();

        await user.click(getByRole(optionListBox, "option", { name: RegExp(mockCadetList[3].firstname) }));
        expect(deregisterField).toHaveValue("");
        expect(deregisterField).toHaveAttribute("aria-expanded", "false");
        expect(optionListBox).not.toBeInTheDocument();

        expect(updateCadetRegistrationForInspection).toHaveBeenCalledWith({
            cadetId: mockCadetList[3].id,
            inspectionId: mockInspectionList[0].id,
            deregister: true
        });
        expect(mutate).toHaveBeenCalled();
    });
    it('should catch exceptions when deregistering a cadet', async () => {
        const { updateCadetRegistrationForInspection } = jest.requireMock("@/dal/inspection");
        const { mutate } = jest.requireMock("@/dataFetcher/inspection").usePlannedInspectionList();
        const { toast } = jest.requireMock("react-toastify");

        const user = userEvent.setup();
        render(
            <DeregistrationOffcanvas
                inspection={mockInspectionList[0]}
                cadetList={mockCadetList}
                onClose={jest.fn()}
            />
        );

        const deregisterFieldGroup = screen.getByTestId("autocomplete-field-group");
        const deregisterField = getByRole(deregisterFieldGroup, "textbox");
        await user.click(deregisterField);

        const optionListBox = getByRole(deregisterFieldGroup, "listbox");
        await user.click(getByRole(optionListBox, "option", { name: RegExp(mockCadetList[3].firstname) }));

        expect(updateCadetRegistrationForInspection).toHaveBeenCalledWith({
            cadetId: mockCadetList[3].id,
            inspectionId: mockInspectionList[0].id,
            deregister: true
        });
        expect(mutate).toHaveBeenCalled();
        expect(mutate).toHaveReturnedTimes(1);
        expect(toast.error).not.toHaveBeenCalled();

        updateCadetRegistrationForInspection.mockRejectedValueOnce(new Error("test error"));
        expect(deregisterField).toHaveAttribute("aria-expanded", "false");
        await user.click(deregisterField);
        expect(deregisterField).toHaveAttribute("aria-expanded", "true");
        await user.click(getByRole(deregisterFieldGroup, "option", { name: RegExp(mockCadetList[4].firstname) }));

        expect(updateCadetRegistrationForInspection).toHaveBeenCalledTimes(2);
        expect(updateCadetRegistrationForInspection).toHaveBeenCalledWith({
            cadetId: mockCadetList[4].id,
            inspectionId: mockInspectionList[0].id,
            deregister: true
        });
        expect(mutate).toHaveReturnedTimes(1);
        expect(toast.error).toHaveBeenCalled();
    });

    it('should be able to remove a cadet from the deregistration list', async () => {
        const { updateCadetRegistrationForInspection } = jest.requireMock("@/dal/inspection");
        const { mutate } = jest.requireMock("@/dataFetcher/inspection").usePlannedInspectionList();
     
        const user = userEvent.setup();
        render(
            <DeregistrationOffcanvas
                inspection={mockInspectionList[0]}
                cadetList={mockCadetList}
                onClose={jest.fn()}
            />
        );

        const removeButtons = screen.getAllByRole("button", { name: /label.remove/i });
        expect(removeButtons).toHaveLength(mockInspectionList[0].deregistrations.length);
        await user.click(removeButtons[0]);
        expect(updateCadetRegistrationForInspection).toHaveBeenCalledWith({
            cadetId: mockInspectionList[0].deregistrations[0].fk_cadet,
            inspectionId: mockInspectionList[0].id,
            deregister: false
        });
        expect(mutate).toHaveBeenCalled();
    });

    it('should catch exceptions when removing a cadet from the deregistration list', async () => {
        const { updateCadetRegistrationForInspection } = jest.requireMock("@/dal/inspection");
        const { mutate } = jest.requireMock("@/dataFetcher/inspection").usePlannedInspectionList();
        const { toast } = jest.requireMock("react-toastify");

        const user = userEvent.setup();
        render(
            <DeregistrationOffcanvas
                inspection={mockInspectionList[0]}
                cadetList={mockCadetList}
                onClose={jest.fn()}
            />
        );

        const removeButtons = screen.getAllByRole("button", { name: /label.remove/i });
        await user.click(removeButtons[0]);
        expect(updateCadetRegistrationForInspection).toHaveBeenCalledWith({
            cadetId: mockInspectionList[0].deregistrations[0].fk_cadet,
            inspectionId: mockInspectionList[0].id,
            deregister: false
        });
        expect(mutate).toHaveBeenCalled();
        expect(mutate).toHaveReturnedTimes(1);
        expect(toast.error).not.toHaveBeenCalled();

        updateCadetRegistrationForInspection.mockRejectedValueOnce(new Error("test error"));
        await user.click(removeButtons[1]);
        expect(updateCadetRegistrationForInspection).toHaveBeenCalledTimes(2);
        expect(updateCadetRegistrationForInspection).toHaveBeenCalledWith({
            cadetId: mockInspectionList[0].deregistrations[1].fk_cadet,
            inspectionId: mockInspectionList[0].id,
            deregister: false
        });
        expect(mutate).toHaveReturnedTimes(1);
        expect(toast.error).toHaveBeenCalled();
    });
});
