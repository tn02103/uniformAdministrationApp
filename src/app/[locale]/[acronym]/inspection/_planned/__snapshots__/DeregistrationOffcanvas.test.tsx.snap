// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeregistrationOffcanvas/> should render 1`] = `
<div
  aria-labelledby="offcanvas-inspection-deregistration"
  aria-modal="true"
  class="offcanvas offcanvas-end offcanvas-toggling show"
  role="dialog"
  style="width: 450px; visibility: visible;"
  tabindex="-1"
>
  <div
    class="offcanvas-header"
  >
    <div
      class="offcanvas-title h5"
      id="offcanvas-inspection-deregistration"
    >
      inspection.planned.deregistration.header
    </div>
    <button
      aria-label="Close"
      class="btn-close"
      type="button"
    />
  </div>
  <div
    class="offcanvas-body"
  >
    <div
      class="row justify-content-between mb-3"
    >
      <div
        class="col-auto"
      >
        <div
          class="position-relativ"
          style="max-width: 200px;"
        >
          <div
            data-testid="autocomplete-field-group"
          >
            <label
              class="form-label"
            >
              inspection.planned.deregistration.label.add
            </label>
            <input
              aria-controls="autocomplete-options"
              aria-expanded="false"
              class="form-control"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <table
      class="table"
    >
      <thead>
        <tr>
          <th />
          <th>
            inspection.planned.deregistration.label.person
          </th>
          <th>
            inspection.planned.deregistration.label.date
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="hoverCol"
        >
          <td>
            <button
              aria-label="inspection.planned.deregistration.label.remove"
              class="hoverColHidden btn btn-light btn-sm"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-x text-danger"
                data-icon="x"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 384 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M376.6 84.5c11.3-13.6 9.5-33.8-4.1-45.1s-33.8-9.5-45.1 4.1L192 206 56.6 43.5C45.3 29.9 25.1 28.1 11.5 39.4S-3.9 70.9 7.4 84.5L150.3 256 7.4 427.5c-11.3 13.6-9.5 33.8 4.1 45.1s33.8 9.5 45.1-4.1L192 306 327.4 468.5c11.3 13.6 31.5 15.4 45.1 4.1s15.4-31.5 4.1-45.1L233.7 256 376.6 84.5z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </td>
          <td>
            Jack
             
            Doe
          </td>
          <td>
            10.05.2025 08:48
          </td>
        </tr>
        <tr
          class="hoverCol"
        >
          <td>
            <button
              aria-label="inspection.planned.deregistration.label.remove"
              class="hoverColHidden btn btn-light btn-sm"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-x text-danger"
                data-icon="x"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 384 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M376.6 84.5c11.3-13.6 9.5-33.8-4.1-45.1s-33.8-9.5-45.1 4.1L192 206 56.6 43.5C45.3 29.9 25.1 28.1 11.5 39.4S-3.9 70.9 7.4 84.5L150.3 256 7.4 427.5c-11.3 13.6-9.5 33.8 4.1 45.1s33.8 9.5 45.1-4.1L192 306 327.4 468.5c11.3 13.6 31.5 15.4 45.1 4.1s15.4-31.5 4.1-45.1L233.7 256 376.6 84.5z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </td>
          <td>
            Jane
             
            Smith
          </td>
          <td>
            30.04.2025 08:48
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
